### ISSUE TITLE
[bazel] Bazel frequently rebuilds the Verilator simulation from scratch

### ISSUE THREAD

**Issue by vogelpi** [lowRISC/opentitan#14353]
When working on chip-level tests I noted that bazel frequently decides to rebuild the Earlgrey Verilator simulation from scratch even without doing hardware changes. For example:
- The other day while waiting for bazel to compile the Verilator simulation I rebased a couple of software-only commits. Once the compilation finished, the test was run. Upon the next command, bazel then decided to rebuild the simulation.
- When continuing work this morning after rebooting my machine, bazel again decided it is necessary to compile the Verilator simulation. Over night, I definitely did zero changes to my OpenTitan source folder.

At the moment, I can work around this, by creating a backup of the built Verilator simulation and then launch it manually. But this behavior is definitely cumbersome as the Verilator build takes a long time and while bazel is busy with that, I cannot launch other commands, e.g., to at least run the test on the FPGA in the meantime.

**Comment by vogelpi** [comment]
Maybe @timothytrippel has an idea of what's going wrong here or a how I could work around this in a better way.

**Comment by timothytrippel** [comment]
At the moment, Bazel builds the Verilator simulation binary by [invoking FuseSoC](https://cs.opensource.google/opentitan/opentitan/+/master:hw/BUILD;drc=00c717be9ae9e3e799c56bd49aadc6b29432f83d;l=24). It should only trigger a rebuild if any dependencies have changed. Unfortunately, the Verilator simulation binary Bazel target [depends on](https://cs.opensource.google/opentitan/opentitan/+/master:hw/BUILD;drc=00c717be9ae9e3e799c56bd49aadc6b29432f83d;l=96) just about all files under `hw/`.  This has the unfortunate benefit of causing the Verilator simulation binary to be rebuilt frequently. 

If you would like to manually avoid this, you can use the `--define DISABLE_VERILATOR_BUILD=true` flag when building a specific target, as described [on the website](https://docs.opentitan.org/doc/ug/bazel_notes/#excluding-verilator-simulation-binary-builds).

> * The other day while waiting for bazel to compile the Verilator simulation I rebased a couple of software-only commits. Once the compilation finished, the test was run. Upon the next command, bazel then decided to rebuild the simulation.

Admittedly, the following behavior is a little peculiar. I would have to see what files were exactly changed to determine if this is expected behavior.

Longer term, it would make sense to put more HW files under Bazel management to take advantage of it's caching, but for now, this is not the case.

**Comment by alphan** [comment]
Looks like this is a long-term enhancement, set to P3, M3. PLMK if you disagree.

**Comment by vogelpi** [comment]
Tim and I had a good discussion on this topic at CHES (thanks @timothytrippel !). To summarize the main discussion points:
- When using `--define DISABLE_VERILATOR_BUILD=true` Verilator simulation targets obviously can't be run. So this cannot be used to avoid rebuilding the Verilator simulation when wanting to rerun a modified test program in Verilator.
- Debugging/fixing the frequent Verilator rebuilding is tricky due to the many dependencies and because of the SV -> C conversion done by Verilator.
- There are huge differences in terms of build time of the Verilator simulation. On my laptop (Ryzen 7) and our workstation (Ryzen 9) it usually takes around 50 minutes. On Tim's workstation it took 5 minutes. That's 10x faster and I agree in that in this case frequent rebuilds are not that bad. But when it takes 50 minutes it's a big blocker.
- Tim discovered that some build arguments were not properly passed from bazel to fusesoc. As a result, the build was only using 1 thread. I think this is tracked in a separate issue.

To enable working around this problem and to also re-enable enabling/disabling tracing at run time, it would be good if bazel would output the Verilator run command including paths to all binaries such that developers can easily make a backup of the Verilator simulation binary and rerun things manually.

**Comment by timothytrippel** [comment]
Thanks for describing the issue @vogelpi ! I filed/linked #15070. I will take a look at these when I get back to California. I think these should be a quick fix.

**Comment by drewmacrae** [comment]
> * The other day while waiting for bazel to compile the Verilator simulation I rebased a couple of software-only commits. Once the compilation finished, the test was run. Upon the next command, bazel then decided to rebuild the simulation.

Could the software changes have affected the test_rom or rom? This can trigger a rebuild.

> * When continuing work this morning after rebooting my machine, bazel again decided it is necessary to compile the Verilator simulation. Over night, I definitely did zero changes to my OpenTitan source folder.

If verilator builds are slow, it's pretty easy to encounter problems with the cache being cleaned of artifacts before we get to reuse them. I use a disk cache to prevent bazel from cleaning my caches before I'm done with it. I added:
```
build --disk_cache=~/bazel_cache
```
to my ~/.bazelrc file and every once in a while I delete the cache by hand.

> To enable working around this problem and to also re-enable enabling/disabling tracing at run time, it would be good if bazel would output the Verilator run command including paths to all binaries such that developers can easily make a backup of the Verilator simulation binary and rerun things manually.

There should be something at the top of the log file. Let me know if it doesn't work for you by filing an issue and assigning it to me.

**Comment by drewmacrae** [comment]
> * Tim discovered that some build arguments were not properly passed from bazel to fusesoc. As a result, the build was only using 1 thread. I think this is tracked in a separate issue.

I think, maybe, the threads argument in #15070 is setting how many threads the binary is compiled to use, not how many threads are used to compile the binary.

**Comment by timothytrippel** [comment]
> > * The other day while waiting for bazel to compile the Verilator simulation I rebased a couple of software-only commits. Once the compilation finished, the test was run. Upon the next command, bazel then decided to rebuild the simulation.
> 
> Could the software changes have affected the test_rom or rom? This can trigger a rebuild.

ROM changes should only retrigger an FPGA bitstream build (though not even if you specify splicing into a prebuilt bitstream from the GCS bucket), not a Verilator sim binary builds. The ROM, flash, and OTP images are passed to the Verilator sim binary at runtime.

**Comment by vogelpi** [comment]
Thanks for your input @drewmacrae , I've now filed a seperate issue regarding the log output and tagged you there.

Regarding the build time. This is a bit weird. Yesterday I was able to build the Verilator simulation binary with a clean cache/from scratch within 5 minutes. Today it takes again 40 min. I did try `bazel clean --expunge` etc. but it doesn't seem to help. Also adding `build --disk_cache=~/bazel_cache` to my bazelrc doesn't seem to help.

**Comment by drewmacrae** [comment]
The 40min verilator builds seem weird but beyond the scope of a [bazel] issue. We might be able to understand the dependencies better by sandboxing Verilator, and could maybe make it more repeatable but as it stands it's hard to speculate about/improve/control. (This is what I meant by #11442). If verilator's argument parser is non-deterministic we might have an explanation, and if that's caused by repeated args then we could fix it in bazel.

Regarding rebuilds. I caught bazel doing it today too. So I can confirm it's happening. I'm not sure if it relates to limitations that prevent us sandboxing Verilator Looks like we didn't fix #9653 after all.

**Comment by drewmacrae** [comment]
I wonder if the build of hw:verilator is scaling to use fewer threads because it's launched in parallel, or if the default for the -j option changed somehow. I'm seeing slow builds now too.

**Comment by msfschaffner** [comment]
CC @milesdai @cfrantz @a-will @timothytrippel 
Is this still a problem today?

**Comment by timothytrippel** [comment]
Unclear, closing for now, but feel free to reopen if the problem arises again.
